### PR TITLE
feat(compiler): perform automatic key insertion in more situations

### PR DESCRIPTION
**Pull request by alicewriteswrongs**
This expands the scope of the automatic key insertion feature that was added in #5143. Previously the feature stopped at the border of any `JsxExpression` syntax node (a bit of arbitrary JavaScript wrapped in curly braces and then inserted into JSX) so that if you were doing something like:

```tsx
<div>{ someCondition && <span>when condition is true</span>}</div>
```

the compiler would insert a key into the `<div>` syntax tree node but _not_ the node for the `<span>`. We did this to just be a bit cautious with the feature because there are a lot of different things that could be going on within a `JsxExpression` node, and some things which could be written within curly braces which would not be safe to transform, such as the following:

```tsx
<div>{ xs.map(x => <span>{ x }</span>) }</div>
```

We wouldn't want to insert a key into the `<span>` syntax tree node because that would result in the dynamically generated `<span>` tags always having the same key attributes.

That said, there are some situations where it is safe to insert a key, such as the `condition || <some-tag />` case shown above. This change adds support for inserting keys in those situations.

STENCIL-1120

## What is the current behavior?

GitHub Issue Number: N/A

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

**Review by christian-bromann — APPROVED** [review]

File: src/compiler/transformers/automatic-key-insertion/index.ts
Comment (rwaskiewicz) [inline src/compiler/transformers/automatic-key-insertion/index.ts:95]: Since we've change this conditional to remove `isJsxExpression` and add `isCallExpression`+`isPropertyAccessExpression`, I think it might make sense to update this comment. I remember you doing this original work, but I think some of the context of this comment is a little lost on me now (TBH, I don't remember what the multiple return statement problem is 😆).

**Review by rwaskiewicz — CHANGES_REQUESTED** [review]
All in all, looks/works good! Just a couple minor asks

File: src/compiler/transformers/automatic-key-insertion/index.ts
Comment (alicewriteswrongs) [reply src/compiler/transformers/automatic-key-insertion/index.ts:95]: yeah, I think that actually the `isPropertyAccessExpression` check isn't necessary overall, I was thinking of situations like

```ts
<div>{ foo.bar(<div />) }</div>
```

where we might not want to transform the `<div />` because we don't know what will happen to it (`.bar` could do anything!) but actually in this case the whole expression `foo.bar(<div />)` is a `CallExpression` where the callee is a `PropertyAccessExpression`. So just checking for `isPropertyAccessExpression` would only guard against situations like

```ts
<div>{ foo.bar }</div>
```

where the property access expression can't have any children that are JSX nodes anyway...

so anyway yeah I think I will remove it and then fix up the comments

Comment (alicewriteswrongs) [reply src/compiler/transformers/automatic-key-insertion/index.ts]: I updated the code here and also fleshed out a comment in the `findRenderMethodVisitor` helper function to explain a bit more the 'multiple return' problem, lmk what you think!

File: src/compiler/transformers/automatic-key-insertion/index.ts
Comment (rwaskiewicz) [inline src/compiler/transformers/automatic-key-insertion/index.ts:97]:
```suggestion
    // re-use the same div element between re-renders, and will just swap out
```

**Review by rwaskiewicz — APPROVED** [review]
LGTM - one non-blocking nit. Comments make sense 🙂

**Comment by rwaskiewicz** [comment]
Actually, @alicewriteswrongs can we hold off on merging this for today? I wanna see if I can get a patch release out to unblock the Lisbon team

**Comment by rwaskiewicz** [comment]
v4.14.1 went out this morning, once this has another ✔️ it should be good to merge. Thanks for holding off on the merge!

**Pull request merged**
